core/src/main/java/org/apache/accumulo/core/data/AbstractId.java

```diff
@@ -14,78 +14,56 @@
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */
-package org.apache.accumulo.core.clientImpl;
-
-import static java.nio.charset.StandardCharsets.UTF_8;
-import static java.util.Objects.requireNonNull;
+package org.apache.accumulo.core.data;
 
 import java.io.Serializable;
 import java.util.Objects;
 
 /**
  * An abstract identifier class for comparing equality of identifiers of the same type.
  */
-public abstract class AbstractId implements Comparable<AbstractId>, Serializable {
+public abstract class AbstractId<T extends AbstractId<T>> implements Comparable<T>, Serializable {
 
-  private static final long serialVersionUID = -155513612834787244L;
+  private static final long serialVersionUID = 1L;
   private final String canonical;
-  private Integer hashCode = null;
 
   protected AbstractId(final String canonical) {
-    requireNonNull(canonical, "canonical cannot be null");
-    this.canonical = canonical;
+    this.canonical = Objects.requireNonNull(canonical, "canonical cannot be null");
   }
 
   /**
-   * The canonical ID
+   * The canonical ID. This is guaranteed to be non-null.
    */
-  public final String canonicalID() {
+  public final String canonical() {
     return canonical;
   }
 
-  public boolean isEmpty() {
-    return canonical.isEmpty();
-  }
-
   /**
    * AbstractID objects are considered equal if, and only if, they are of the same type and have the
    * same canonical identifier.
    */
   @Override
   public boolean equals(final Object obj) {
-    if (this == obj)
-      return true;
-    return obj != null && Objects.equals(getClass(), obj.getClass())
-        && Objects.equals(canonicalID(), ((AbstractId) obj).canonicalID());
+    return this == obj || (obj != null && Objects.equals(getClass(), obj.getClass())
+        && Objects.equals(canonical(), ((AbstractId<?>) obj).canonical()));
   }
 
   @Override
   public int hashCode() {
-    if (hashCode == null) {
-      hashCode = Objects.hash(canonicalID());
-    }
-    return hashCode;
+    return canonical().hashCode();
   }
 
   /**
    * Returns a string of the canonical ID
    */
   @Override
   public String toString() {
-    return canonical;
-  }
-
-  /**
-   * Return a UTF_8 byte[] of the canonical ID.
-   */
-  public final byte[] getUtf8() {
-    return canonical.getBytes(UTF_8);
+    return canonical();
   }
 
   @Override
-  public int compareTo(AbstractId id) {
-    requireNonNull(id, "id cannot be null");
-    return this.canonicalID().compareTo(id.canonicalID());
+  public int compareTo(T other) {
```

Comment:
Can `compareTo()` and `equals()` disagree on the equality of some objects?

Reply:
Not in any normal situation. equals checks type and canonical field. compareTo compares whatever the generic parameter type is. So, in order for that to happen, one would have to do something like `public class NewIdType extends AbstractId<TableId>` or subclass TableId, and not override equals.

Reply:
Ok I see, so the following would fail to compile because T is of different type.

```java
NamespaceId id1 = new NamespaceId.of("1");
TableId id2 = new TableId.of("1");
int c = id1.compareTo(id2);
```

Also I suppose the following would fail because T is not defined?

```java
AbstractId id1 = new NamespaceId.of("1");
AbstractId id2 = new TableId.of("1");
int c = id1.compareTo(id2);
```

Reply:
Yes for the first case. I haven't tried the second case, but it probably assumes `T` is `AbstractId<?>`. But, if you're already willing to accept the lack of type checking rigor and ignore all the rawtypes and unchecked cast warnings from the compiler when you do something like that, then you probably don't care that the comparison isn't rigorous either.

If we really wanted to do the work of the compiler-time type checking in `compareTo`, we could add a type check like we have in the `equals` method, but I don't see the point of that when the only time it matters is when the programmer themselves don't care about compiler type checking rigor.

The weirder, and less obvious, case would be when the programmer creates a new type, related to one of the existing ID types, like:

```java
// or just subclass TableId itself
public class MyNewIdType extends AbstractId<TableId> { ... }

int exampleMethod() {
  TableId id1 = TableId.of("1");
  MyNewIdType id2 = new MyNewIdType("2");
  return id2.compareTo(id2);
}
```

But, in that case, the programmer is explicitly establishing a relationship between their type and the TableId type, so it's probably fine. Weird, but fine. And, in any case, whenever you subclass something that's Comparable, you're going to have to put thought into what that means anyway. That's just a general thing to consider with Comparable stuffs, and not unique to this scenario in any way.